MI-219 Local SQLite Database

limphilip opened the pull request:
This pull request contains the updates to include a local SQLite database to store some persistent structured data, such as the user profile information. Here is the summary of the changes:

- The `Cache` package now contains two new sub-packages named `LocalDatabase` and `LocalDataGateways`. The cache logic classes (`UserCache` for now) use their corresponding `LocalDataGateway` classes in order to retrieve information in the SQLite storage. The `LocalDatabase` package is responsible for creating the database and its table when necessary, and defining the path/location of the database, used by the local gateways.
- New files related to SQLite are also added through installing the SQLite NET nuget package.
- The `User` model has been updated, because the `LocalDatabase` package bases itself on the `Model` package to define the database table. I.e. the `Models` are the database schemas.
- Unit tests for the local gateways have been added to test the following functionalities: Add, Update, Find, Delete a user.
- Dependencies to the `sqlite net nuget package` is added.

Note: also related to #47

limphilip committed:
MI-212 Initial implementation of the local SQLite database + unit tests

This initial implementation updates the project with the SQLite
component. It allows the application to store structured objects in the
a local database using object relational model.

The Cache package now contains a LocalDatabase package that defines the
location of the database, and that is responsible for creating the
tables when missing. The Cache also contains a new package named
LocalDataGateways. This new package contains gateways responsible for
communicating with the local database, the package is used by the logic
classes such as UserCache. The UserMapper will eventually communicate
with this UserCache in order to access the database indirectly, but this
is not implemented yet.

For the database to know how to define the table, the User model has
been updated with a primary key, and all the lists/relations with other
models have been ignored due to size limitation.

The SQLite and SQLiteAsync are classes added by the components handling
the SQLite database.

Unit tests for the UserLocalDataGateway has been added to test the
following functionalities: adding a user, deleting a user, finding a
user by id, updating a user. All four cases passed.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/MI-220/FileCaching' into MI-219/LocalSqliteDB

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/MI-220/FileCaching' into MI-219/LocalSqliteDB

limphilip committed:
MI-212 Updated UserMapper to look in the user cache for the user before asking the server

The UserMapper.GetUser() will now look into the cache first to find a
user, before asking the server.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-219/LocalSqliteDB

limphilip committed:
Remove tag for ChatAdapter in the CSPROJ file since this file has been replaced.

During the merge, the file has been replaced, but the old tag was still
kept. It is now removed.

limphilip committed:
MI-213 Updated test cases with more compact and meaningful assertions

For readability and easiness to pin point a bug, assertions have been
compacted.

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-219/LocalSqliteDB

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-219/LocalSqliteDB

limphilip committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into MI-219/LocalSqliteDB

Ideawin commented on UnitTests/CacheTest/LocalDataGatewayTest/UserLocalDataGatewayTest.cs:
```diff
@@ -0,0 +1,281 @@
+            User user = new User
+            {
+                id = userId,
+                facebookID = userId,
```
Maybe you should use two different ids for userId and facebookID in your tests to make sure that it saves the right thing for both

limphilip replied:
Thanks for pointing this out @Ideawin ! The facebook IDs have now been changed to be different from the user ID in order to ensure that they are both stored/retrieved correctly.

limphilip committed:
MI-213 Added facebookd ID that is different from the user ID

The facebook ID used in the test were the same as the user ID. Different
ID numbers are now used for the faceobook ID to ensure that both the
user ID and the facebook ID are stored correctly.

Ideawin commented:
It's perfect and complete with unit tests. Merging
